### ISSUE TITLE
How do I read multiple branches with jagged structure through ```uproot.iterate()```

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/uproot4#433]
### Discussed in https://github.com/scikit-hep/uproot4/discussions/430

<div type='discussions-op-text'>

<sup>Originally posted by **MoAly98** September  1, 2021</sup>
Assume I have 3 branches in an input file called ```file``` in a tree ```tree```:
```branch1: type = AsJagged(AsDtype('>f4'))```
```branch2: type = AsJagged(AsDtype('>f4'))```
```branch3: type = AsDtype('>f4')```
My aim is to read in these 3 branches using uproot and save them into a dataframe. If I run 
```
for batch in uproot.iterate(file:tree, filter_name=[branch1, branch2, branch3], library='pd'):
      print(batch)
```

I run into the following error from the iterate method call:
```
  File "/afs/cern.ch/user/m/maly/thbb_env/lib/python3.7/site-packages/uproot/behaviors/TBranch.py", line 226, in iterate
    arrays = library.global_index(item, global_offset)
  File "/afs/cern.ch/user/m/maly/thbb_env/lib/python3.7/site-packages/uproot/interpretation/library.py", line 1084, in global_index
    index = arrays.index.values  # pandas<0.24.0
AttributeError: 'builtin_function_or_method' object has no attribute 'values'

```

This problem is not present when I read the same 3 branches with 
```
uproot.open(file:tree).arrays(filter_name=[branch1, branch2, branch3], library = 'pd' )
```
, which yields a tuple of dataframes with length = number of jagged arrays being read as expected (which is 2 here). 
I can't see anything in the ```uproot``` documentation which prohibiits reading multiple jagged arrays into pandas dataframes via the ```iterate``` method, so I'm not sure what is causing the above error. 

I am using version ```4.0.0``` of ```uproot```. </div>

**Comment by jpivarski** [comment]
@MoAly98 Moved Discussion to Issue because it's a bug-report.

**Comment by jpivarski** [comment]
The key thing that was happening here is that branch1 and branch2 have different numbers of values in each entry ("_differently_ jagged"), so they can't be in the same DataFrame, so `iterate` was yielding _tuples_ of DataFrames rather than just DataFrames. When you iterated over more than one file, it tried to set the index of all files after the first one to globally increasing numbers (so that indexes would not go back to zero just because you're opening a new file), and Uproot's `global_index` function assumes that it gets a DataFrame, rather than a tuple of DataFrames. That last assumption is a bug and it's fixed with

https://github.com/scikit-hep/uproot4/blob/0bc31f3729b1876814e05261698baa0584d9c207/src/uproot/interpretation/library.py#L1087-L1089

in PR #432.

Since the error message pointed to a line in `global_index` with a comment about `pandas<0.24.0`, I tested it with Pandas 0.22.0 and found another issue: branch3 would come up all NaN because of an old behavior in Pandas. I added a comment to the codebase for future generations, but this seems like a really old version of Pandas and I think we shouldn't be supporting anything that old. Still, I don't know when it was fixed in Pandas to put a Pandas version dependency in Uproot. The Pandas version you're using is likely to be able to put branch3 in your DataFrames without NaN.

Uproot 4.0.0 is also rather old: December 5, 2020 if I remember right (because it was an auspicious version number update). On Monday, 4.1.1 was released, and the bug-fix in PR #432 will go into an Uproot 4.1.2 soon.

If you need this working without waiting for a new release, you could have `uproot.iterate` yield Awkward Arrays with `library="ak"` and call [ak.to_pandas](https://awkward-array.readthedocs.io/en/latest/_auto/ak.to_pandas.html) to make your DataFrames.

**Comment by MoAly98** [comment]
Hi @jpivarski,

Thanks a lot for the quick and helpful reply. I should add that I see the same error when I pass just one file rather than a list of files. I guess you mean when a new chunk of data is being processed then the indexing causes trouble?

Unfortunately the ```ak.to_pandas()```method is not doing what I need (which is another story that probably belongs in the ```awkward``` Q&A). It could be useful to this discussion that I outline this issue:
What I would like to achieve is a structure like (shown for 1 event):

|       |           |           | Branch1 | Branch2 | Branch3 |
|:-----:|:---------:|:---------:|:-------:|:-------:|:-------:|
| entry | subentry1 | subentry2 |         |         |         |
|   0   |     0     |     0     |    X0   |    Y0   |    Z    |
|       |           |     1     |    X0   |    Y1   |    Z    |
|       |           |     2     |    X0   |    Y2   |    Z    |
|       |           |     3     |    X0   |    Y3   |    Z    |
|       |     1     |     0     |    X1   |    Y0   |    Z    |
|       |           |     1     |    X1   |    Y1   |    Z    |
|       |           |     2     |    X1   |    Y2   |    Z    |
|       |           |     3     |    X1  |    Y3   |    Z    |

where ```subentryX``` is the index for ```BranchX ```. However, ```ak.to_pandas()```  seem to be collapsing the indices such that I get:
|       |          | Branch1 | Branch2 | Branch3 |
|-------|----------|---------|---------|---------|
| entry | subentry |         |         |         |
| 0     | 0        | X0      | Y0      | Z       |
|       | 1        | X1      | Y1      | Z       |
|       | 2        | X2      | Y2      | Z       |
|       | 3        | X3      | Y3      | Z       |

and I can't seem to achieve the structure I want from playing around with the ```how``` argument in ```ak.to_pandas()```. 

I'm not sure if I'm doing someting wrong, or if this is a bug in the way ```uproot.iterate()``` is saving the information in the awkward arrays causing the loss of information.

**Comment by jpivarski** [comment]
You'll only get a table like the first one if the data were doubly jagged, like a `std::vector<std::vector<float>>` for each entry. You said that branch1 and branch2 both have interpretation `AsJagged(AsDtype('>f4'))`, which is what you'd get with `std::vector<float>` data (or variable-length arrays: there are multiple C++ types that lead to the same interpretation).

But if branch1 and branch2 are "_differently_ jagged," they have different numbers of values in each entry, then naively they can't go into the same DataFrame. You can force them into the same DataFrame by specifying `how="inner"`, `"outer"`, etc., which tells [pd.merge](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html) how to match subentry indexes of branch1 with subentry indexes of branch2, because there is a different number of them.

If I'm wrong in my interpretation that branch1 and branch2 are differently jagged, then I don't know how you got the error. I was only able to reproduce it with differently jagged data, because only then do you get a tuple of DataFrames instead of just one DataFrame, and that's what the error message was saying (obliquely: `index` is a method for tuples and a property for DataFrames, and it was complaining that a method does not have an attribute `arrays`, which a DataFrame index does).

**Comment by MoAly98** [comment]
You were right in your interpretation -- as a concrete example, I'm trying to read in ``leptons_pt ```and ```jets_pt ``` branches which are variable-length arrays for each event. For such case, I would like to retain for each event values of pt for all jets as well as pt of all leptons.  I am able to generate the first table using some pandas ```join``` method manipulation using the dataframes tuple, but understandably ```ak.to_pandas()``` is not able to do this for me. 

Is there an indication on when we can expect v4.1.2 ? 

Thanks again for all the amazing work on ```uproot``` and other packages for pyhep !

**Comment by jpivarski** [comment]
The `how` parameter of [uproot.iterate](https://uproot.readthedocs.io/en/latest/uproot.behaviors.TBranch.iterate.html) is passed to Pandas for merging, but if it has to be done in a separate line of code, I don't think that's a bad thing. If you want to be very careful about this, so that a chunk of events that _happen to have_ the same number of leptons as jets (for all events in that chunk) don't get accidentally merged into the same DataFrame, you could do two `uproot.iterate` calls, set the `step_size` of both to an exact number of entries (as opposed to an approximate number of bytes), and then use Python's `zip` to walk over both of them at the same time.

```python
events = uproot.iterate(files, ["event_branches", ...], step_size=10000)
leptons = uproot.iterate(files, ["lepton_branches", ...], step_size=10000)
jets = uproot.iterate(files, ["jet_branches", ...], step_size=10000)

for df_event, df_lep, df_jet in zip(events, leptons, jets):
    # carefully merge the DataFrames in a way that won't break in extreme cases
```

Incidentally, this would be a way of doing a "friend TTree" analysis, too.

Uproot releases don't have a fixed schedule, so if you need PR #432 to be in a fixed release, test it for your use-case in main ([like this](https://adamj.eu/tech/2019/03/11/pip-install-from-a-git-repository/) or clone the GitHub repo and do `pip install .` from its directory) and if it works, I'll make the release.

**Comment by MoAly98** [comment]
AHA! I hadn't thought of this extreme case so thanks a lot for pointing it out :)

I tested the bug-fix and it seems to solve my issue of processing multiple jagged arrays with ```uproot.iterate()```. Thanks a lot !
